File: ACTransit.RestroomFinder.Native/Android/RestroomFinder/app/src/main/java/org/actransit/restroomfinder/Infrastructure/AppStorage.java
package org.actransit.restroomfinder.Infrastructure;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.provider.Settings;
import androidx.core.app.ActivityCompat;
import android.telephony.TelephonyManager;

import com.tellexperience.jsonhelper.JsonMapper;

import org.actransit.restroomfinder.LoginActivity;
import org.actransit.restroomfinder.Model.KeyValue;
import org.json.JSONObject;

import java.util.Calendar;
import java.util.UUID;


/**
 * Created by DevTeam on 6/21/16.
 */
public class AppStorage extends Activity{
    private static final String PREFS_NAME = "RestroomFinderFile";
    private static final String disclaimerKey = "Disclaimer_Agreed";
    //private static final String disclaimerSessionKey = "Disclaimer_Agreed_SessionId";
    private static final String disclaimerPopupKey = "DisclaimerPopup_Agreed";


    private static final String badgeKey = "badge";
    private static final String phoneNumberKey = "phoneNumber";
    private static final String firstNameKey = "firstName";
    private static final String lastNameKey = "lastName";
    private static final String sessionKey = "SessionId";


    private static SharedPreferences settings;
    private static AppStorage current;
    private AppStorage(Context context){
        settings= context.getSharedPreferences(PREFS_NAME, 0);
    }

    private KeyValue badge;
    private KeyValue sessionId;
    private KeyValue firstName;
    private KeyValue lastName;
    private KeyValue phoneNumber;

    public static AppStorage Current(Context context){
        if (current!=null)
            return current;
        current=new AppStorage(context);
        return current;
    }

    public boolean shouldPopupDisclaimer() {
        String value=settings.getString(disclaimerPopupKey,"");
        if (value.isEmpty())
            return true;
        try{
            JSONObject jObj=new JSONObject(value);
            KeyValue keyValue= JsonMapper.toObject(jObj, KeyValue.class);
            Calendar now=Calendar.getInstance();
            Calendar c= Calendar.getInstance();
            c.setTime(keyValue.updateDate);
            return c.get(Calendar.YEAR) != now.get(Calendar.YEAR) ||
                    c.get(Calendar.MONTH) != now.get(Calendar.MONTH) ||
                    c.get(Calendar.DAY_OF_MONTH) != now.get(Calendar.DAY_OF_MONTH);

        }
        catch (Exception e){
            e.printStackTrace();
            return true;
        }
    }

    public void setPopupDisclaimer(){
        saveItem(disclaimerPopupKey,"set");
    }


//    public boolean isFirstTimeRunningApplication(){
//        String disClaimerAgreed = settings.getString(disclaimerKey, "");
//        return disClaimerAgreed=="";// | true;
//    }

//    public void setFirstTimeRunningApplication(String badge, String sessionId) {
//        saveItem(disclaimerKey,badge);
//        saveItem(sessionKey,sessionId);
//    }

    public Boolean isEmpty(){
        return badge == null &&
            sessionId==null &&
            firstName == null &&
            lastName == null &&
            phoneNumber== null;
    }

    public void setDisclaimerShown() {
        saveItem(disclaimerKey,"1");
    }


    public void setCurrentLoginInformation(String firstName, String lastName, String badge, String phoneNumber, String sessionId) {
        saveItem(badgeKey,badge);
        saveItem(firstNameKey,firstName);
        saveItem(lastNameKey,lastName);
        saveItem(phoneNumberKey,phoneNumber);
        saveItem(sessionKey,sessionId);

    }
    public void logOut()  {
        Common.Loggedin=false;
        saveItem(disclaimerKey,"");
        saveItem(badgeKey,"");
        saveItem(firstNameKey,"");
        saveItem(lastNameKey,"");
        saveItem(phoneNumberKey,"");
        saveItem(sessionKey,"");

        invalidateCache();

        MyApplication.NavigateTo(LoginActivity.class);
        //let _=Common.NavigateTo("Login")
    }

    public String getBadge(){
        if (badge==null)
            badge=restoreItem(badgeKey);
        if (badge==null)
            return null;
        return badge.value;
    }
    public String getSessionId(){
        if (sessionId==null)
            sessionId=restoreItem(sessionKey);
        if (sessionId==null)
            return null;
        return sessionId.value;
    }
    public String getFirstName(){
        if (firstName==null)
            firstName=restoreItem(firstNameKey);
        if (firstName==null)
            return null;
        return firstName.value;
    }
    public String getLastName(){
        if (lastName==null)
            lastName=restoreItem(lastNameKey);
        if (lastName==null)
            return null;
        return lastName.value;
    }
    public String getPhoneNumber(){
        if (phoneNumber==null)
            phoneNumber=restoreItem(phoneNumberKey);
        if (phoneNumber==null)
            return null;
        return phoneNumber.value;
    }

    public Boolean isDisclaimerShown() {
        KeyValue disclaimer=restoreItem(disclaimerKey);
        boolean tempResult=(disclaimer==null || !disclaimer.value.equals("1") );
        boolean result = !tempResult;
        return result;
    }



    public void saveItem(String key,String value){
        try{
            invalidateCache();
            JSONObject jObj=new JSONObject();
            jObj.put("key",key);
            jObj.put("value",value);
            jObj.put("updateDate",Calendar.getInstance().getTime().getTime());
            String v = jObj.toString();
            SharedPreferences.Editor editor = settings.edit();
            editor.putString(key, v);
            editor.commit();

        }
        catch(Exception e){
            e.printStackTrace();
        }
    }
    public KeyValue restoreItem(String key) {
        String value=settings.getString(key,"");
        if (value.isEmpty())
            return null;
        try{
            JSONObject jObj=new JSONObject(value);
            KeyValue keyValue=JsonMapper.toObject(jObj, KeyValue.class);
            return keyValue;
        }
        catch (Exception e){
            e.printStackTrace();
        }
        return null;
    }

    private void invalidateCache(){

        badge=null;
        sessionId=null;
        firstName = null;
        lastName = null;
        phoneNumber= null;
    }

}

File: ACTransit.RestroomFinder.Native/Apple/Restroom/AppStorage.swift
//
//  AppStorage.swift
//  Restroom
//
//  Created by DevTeam on 2/22/16.
//  Copyright © 2016 DevTeam. All rights reserved.
//

import Foundation
import SharedFramework
class AppStorage{
    static let disclaimerKey = "Disclaimer_Agreed"
    static let badgeKey = "badge"
    static let phoneNumberKey = "phoneNumber"
    static let firstNameKey = "firstName"
    static let lastNameKey = "lastName"
    static let disclaimerPopupKey = "DisclaimerPopup_Agreed"
    static let sessionKey = "SessionId"
    
    static var _keyValues: [KeyValue]!
    
    static func shouldPopupDisclaimer() -> Bool{
        let keyvalue = keyValues?.tryGetValue(disclaimerPopupKey)
        
        if (keyvalue == nil){
            return true
        }
        else {
            //print ("now: \(DateTime.now().toString()) vs \(keyvalue?.updateDate?.toString())")
            let now = DateTime.now()
            return (keyvalue?.updateDate?.year != now.year) ||
                    (keyvalue?.updateDate?.month != now.month) ||
                    (keyvalue?.updateDate?.day != now.day)
        }
    }

    static func setPopupDisclaimer() throws{
        try saveValue(key: disclaimerPopupKey, value: "set", dateTime: DateTime.now())
    }
    
    
    static func isEmpty() -> Bool{
        //return true
        if (keyValues == nil){
            return true
        }
        return false
    }
    
    static func setCurrentLoginInformation(_ firstName: String, lastName: String, badge: String, phoneNumber: String, sessionId: String) throws {
        
        try saveValue(key: badgeKey, value: badge, dateTime: DateTime.now())
        try saveValue(key: firstNameKey, value: firstName, dateTime: DateTime.now())
        try saveValue(key: lastNameKey, value: lastName, dateTime: DateTime.now())
        try saveValue(key: phoneNumberKey, value: phoneNumber, dateTime: DateTime.now())
        try saveValue(key: sessionKey, value: sessionId, dateTime: DateTime.now())
    }
    
    static func setDisclaimerShown() throws {
        try saveValue(key: disclaimerKey, value: "1", dateTime: DateTime.now())
    }
    
    
    static func saveValue(key:String, value:String?, dateTime:DateTime?) throws{
        var res = keyValues
        if (res == nil){
            res = [KeyValue]()
        }
        let keyvalue = res?.tryGetValue(key)
        if (keyvalue == nil){
            res?.append(KeyValue(key: key, value: value, updateDate: dateTime))
        }
        else{
            keyvalue?.updateDate = DateTime.now()
            keyvalue?.value = value
        }
        invalidateCache()
        _=KeyValue.Save(res!)
    }
    
    static func logOut() throws {
        try saveValue(key: disclaimerKey, value: "", dateTime: DateTime.now())
        try saveValue(key: badgeKey, value: "", dateTime: DateTime.now())
        try saveValue(key: firstNameKey, value: "", dateTime: DateTime.now())
        try saveValue(key: lastNameKey, value: "", dateTime: DateTime.now())
        try saveValue(key: phoneNumberKey, value: "", dateTime: DateTime.now())
        try saveValue(key: sessionKey, value: "", dateTime: DateTime.now())
        
        invalidateCache()
//        _=Common.NavigateToModal("Disclaimer", modalCompletion: { (controller) in
//        })
        
        let _=Common.NavigateTo("Login")
    }
    static var badge : String {
        get{
            return keyValues?.tryGetValue(badgeKey)?.value ?? ""
        }
    }
    static var phoneNumber : String {
        get{
            return keyValues?.tryGetValue(phoneNumberKey)?.value ?? ""
        }
    }
    static var firstName : String {
        get{
            return keyValues?.tryGetValue(firstNameKey)?.value ?? ""
        }
    }
    static var lastName : String {
        get{
            return keyValues?.tryGetValue(lastNameKey)?.value ?? ""
        }
    }
    static var sessionId : String {
        get{
            return keyValues?.tryGetValue(sessionKey)?.value ?? ""
        }
    }
    
    static var disclaimerShown: Bool {
        get{
            let value=keyValues?.tryGetValue(disclaimerKey)?.value
            return !(value==nil || value != "1")
        }
    }

    
    fileprivate static var keyValues: [KeyValue]? {
        get{
            if (_keyValues == nil){
                _keyValues = KeyValue.Load()
            }
            return _keyValues
        }
    }
    fileprivate static func invalidateCache(){
        _keyValues = nil;
    }
}
